### PR TITLE
feat: allow users to add insights to multiple dashboards

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

stacked on top of #9524 this adds UI to allow testing addition of insights to multiple dashboards

Design at https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=8803%3A44524

## Changes

multidash-examples

Adds a new modal behind a feature flag that allows you to

* see which dashboards an insight is on
* search a list of dashboards
* remove insight from dashboards
* add insight to a dashboard
* create a new dashboard (so an insight can be added to it)

NB while this is safe to merge behind a flag. We likely need to add a way to have more than one `last_refresh` time for an insight before we can start using this widely

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

running it locally and using it to add and remove dashboards

### CODE REVIEW

**Comment by pauldambra** [comment]
Add to dashboard button

Screenshot 2022-04-26 at 17 00 35

Screenshot 2022-04-26 at 17 00 18

Screenshot 2022-04-26 at 17 00 56

**Comment by pauldambra** [comment]
add to dashboard modal

Screenshot 2022-04-26 at 17 03 22

**Comment by pauldambra** [comment]
tagging @lottiecoxon and @clarkus for review since the UI spans their handover :)

**Review by clarkus — CHANGES_REQUESTED** [review]
Looks great. The one opportunity I see is to automatically add the insight to a newly created dashboard. I noticed in your gif that you created the dashboard, it appended to the list, you clicked to add the insight. Can we collapse all that into the submit action for adding a new dashboard? For example:

1. Click to create a new dashboard
2. Name the dashboard and submit
3. See the new dashboard added to the list and shown as `added` 

Thoughts?

**Comment by pauldambra** [comment]
> Looks great. The one opportunity I see is to automatically add the insight to a newly created dashboard. I noticed in your gif that you created the dashboard, it appended to the list, you clicked to add the insight. Can we collapse all that into the submit action for adding a new dashboard? For example:
> 
> 1. Click to create a new dashboard
> 2. Name the dashboard and submit
> 3. See the new dashboard added to the list and shown as `added`
> 
> Thoughts?

Yep, that would be nicer! The current flow is as implemented here... create the dashboard and then have to click to save it

I just had a quick run at implementing the nicer flow and there's a jumble of interacting logics that get very upset... Since this is behind a flag, ideally this could be a follow-up - I'll take your guide on whether we should block here though

**Comment by clarkus** [comment]
Not a blocker, but definitely something we should track for improvement. Are the logic constraints mostly tied to components we're reusing or is there anything else I can keep in mind when working on this next time?

**Comment by pauldambra** [comment]
> Are the logic constraints mostly tied to components we're reusing or is there anything else I can keep in mind when working on this next time?

@clarkus it's the interactions of kea logics... I'd really like to work on this, not just because it's a better experience but to grab @mariusandra and level up my knowledge. It's definitely fixable!

So, a technical not design constraint :)

File: frontend/src/lib/components/SaveToDashboard/AddToDashboard.scss [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+    .ant-modal-title {
+        font-weight: 600;
+        font-size: 11px;
+        line-height: 24px;
+        letter-spacing: 0.5px;
+        text-transform: uppercase;
+        color: var(--text-muted);
+    }
+}
```
Comment (mariusandra) [inline frontend/src/lib/components/SaveToDashboard/AddToDashboard.scss:53]: Going to tag @Twixes here for lemon input.

File: frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx [hunk 2]
```diff
@@ -4,21 +4,143 @@ import { Select, Modal } from 'antd'
...
+            {
+                ...insight,
+                dashboards: (insight.dashboards || []).filter((d) => d !== dashboardToRemove),
+            },
+            () => {
+                reportRemovedInsightFromDashboard()
+                lemonToast.success('Insight removed from dashboard')
+            }
+        )
+    }
```
Comment (mariusandra) [inline frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx:66]: Can these functions be moved to `saveToDashboardModalLogic`? Ideally we want to move as much data manipulation complexity into kea as possible.

File: frontend/src/lib/components/icons.stories.tsx [hunk 3]
```diff
@@ -59,3 +60,55 @@ export function Icons(): JSX.Element {
         />
     )
 }
+
+export function IconWithCountBubble(): JSX.Element {
+    const betweenOneAndTwenty = Math.floor(Math.random() * (20 - 1 + 1)) + 1
```
Comment (mariusandra) [inline frontend/src/lib/components/icons.stories.tsx:65]: Can you remove the randomness?

Theoretically this breaks the component's purity, which React assumes we keep. Now every render renders something different, and it's not guaranteed when anything will be rendered.

Plus it'll be another thing to fix, once we introduce visual regression testing.

**Review by mariusandra — COMMENTED** [review]
Looks really good! 👍  Some comments inline and below to get it over the line. Will also tag @Twixes to feedback the lemons.

1. Could we make this popup longer if you have a huge screen? It's weird to scroll for dashboards when there's all that space available.

image

2. Can the number be capped at 99 max? There will be someone who will add an insight to 101 dashboards. Just `Math.max` will do, nothing special needed here. If there's space, "99+" with a smaller font would be even better.
image

File: frontend/src/lib/components/icons.tsx [hunk 4]
```diff
@@ -1,7 +1,21 @@
...
 import './icons.scss'
 
+interface IconWithCountProps {
+    count: number
+    showZero?: boolean
+}
+
+export function IconWithCount({ count, showZero, children }: PropsWithChildren<IconWithCountProps>): JSX.Element {
+    return (
+        <div style={{ position: 'relative' }}>
```
Comment (Twixes) [inline frontend/src/lib/components/icons.tsx:12]: The sizing of this seems a bit off: it's 24x25px in "Add to dashboard" and in Storybook it looks like weirdly like this:

Screen Shot 2022-04-28 at 13 59 51

`IconWithCount` should be exactly the same size as the underlying icon (so 24x24px by default, or 1em by 1em).

File: frontend/src/lib/components/icons.scss [hunk 5]
```diff
@@ -22,3 +21,19 @@
...
+.icon-count-bubble {
+    position: absolute;
+    width: 16px;
+    height: 16px;
+    line-height: 16px;
+    top: -0.3rem;
+    right: -0.3rem;
+    color: var(--white);
+    background: var(--primary);
+    font-size: 0.6rem;
```
Comment (Twixes) [inline frontend/src/lib/components/icons.scss:34]: These values are a bit off in px terms, as 0.6rem is typically 9.6px (unless user changes root font size from 16px). They should generally be multiples of 1/16.
```suggestion
    font-size: 0.5625rem;
```

File: frontend/src/lib/components/icons.scss [hunk 6]
```diff
@@ -22,3 +21,19 @@
...
     }
 }
+
+.icon-count-bubble {
+    position: absolute;
+    width: 16px;
+    height: 16px;
+    line-height: 16px;
+    top: -0.3rem;
+    right: -0.3rem;
```
Comment (Twixes) [inline frontend/src/lib/components/icons.scss:31]: The count is a bit further from the center of the icon in the design as the icon is more easily recognizable then.
```suggestion
    top: -0.375rem;
    right: -0.375rem;
```

**Review by Twixes — CHANGES_REQUESTED** [review]
Love the feature! Left some lemon (and general code+UX) input

File: frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx (hunk 2 above)
Comment (pauldambra) [reply frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx:66]: Hmm, that's what the logics got upset about when I tried to have the new dashboard auto selected 🤔

**Comment by pauldambra** [comment]
~~I think I've addressed all of the feedback except the move of the functions into the logic.~~

~~I'd like to get this merged (since it's behind a flag) so I can test the many-to-many insights in prod properly and handle the location of the functions as a follow-up~~

Paired with @mariusandra  and learned a lot of kea! 🙌

File: frontend/src/lib/utils.test.ts [hunk 7]
```diff
@@ -189,7 +189,7 @@ describe('pluralize()', () => {
         expect(pluralize(1, 'word', undefined, false)).toEqual('word')
     })
     it('handles plural cases', () => {
-        expect(pluralize(28321, 'member')).toEqual('28321 members')
+        expect(pluralize(28321, 'member')).toEqual('28 321 members')
```
Comment (mariusandra) [inline frontend/src/lib/utils.test.ts:192]: So we have 28 321-member units?

**Review by mariusandra — CHANGES_REQUESTED** [review]
The dashboard part looks good, but what is all the other stuff in this PR, regarding removing a feature flag, fixing the person modal, updating a kafka queue test, etc? 😕

**Comment by pauldambra** [comment]
> The dashboard part looks good, but what is all the other stuff in this PR,

@mariusandra and there was me saying I was enjoying stacked PRs 🤦 

needed to merge from master locally... just the changes from this PR showing now

File: frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx [hunk 8]
```diff
@@ -1,24 +1,151 @@
...
+
+export function AddToDashboardModal({ visible, closeModal, insight }: SaveToDashboardModalProps): JSX.Element {
+    const logic = saveToDashboardModalLogic({
+        insight: insight,
+        fromDashboard: insight.dashboards?.[0] || undefined,
+    })
+
+    const { searchQuery, currentDashboards, orderedDashboards, scrollIndex } = useValues(logic)
+    const { setSearchQuery } = useActions(logic)
+    const { addNewDashboard } = useActions(logic)
```
Comment (mariusandra) [inline frontend/src/lib/components/SaveToDashboard/SaveToDashboardModal.tsx:77]:
```suggestion
    const { setSearchQuery, addNewDashboard } = useActions(logic)
```

**Review by mariusandra — APPROVED** [review]
Still not sure why these buttons aren't the same high, the rest LGTM:

image

**Comment by pauldambra** [comment]
> Still not sure why these buttons aren't the same high

it's coming from [the LemonModal section styling](https://github.com/PostHog/posthog/blob/master/frontend/src/lib/components/LemonModal/LemonModal.scss#L36)

I've added a workaround for now, and will bother @Twixes next week about it :)

**Comment by pauldambra** [comment]
@clarkus @lottiecoxon I took advantage of @mariusandra brain and now when you add the new dashboard we automagically add the insight to it 🎉

**Pull request merged**